### DISCUSSION TITLE
RFC: Multi-Parameter Type Classes

### DISCUSSION THREAD

**Discussion by eudoxia0** [austral/austral#418] [Ideas]
Multi-Parameter Type Classes are exactly what it says on the tin: type classes with more than one parameter. As is often the case, these increase the expressive power a great deal. However, they come at the cost that the rules for ambiguity-prevention are more complex.

# Example

A simple, if somewhat uninspiring, example:

```
typeclass Isomorphism(A: Type, B: Type) is
    method forward(a: A): B;

    method backward(b: B): A;
end;
```

Arithmetic could be made more general, such that the left and right hand side of an arithmetic operator can be different types, for example (this example uses [associated types][assoc]):

[assoc]: https://github.com/austral/austral/discussions/417

```
typeclass Arithmetic(A: Type, B: Type) is
    associated type R: Type;

    method add(lhs: A, rhs: B): R;

    -- etc.
end;

instance Arithmetic(Scalar, Matrix) is
    associated type R := Matrix;

    method add(lhs: Scalar, rhs: Matrix): Matrix is
        -- etc.
    end;
end;
```

# Unknowns

I feel like this feature _can_ be a net improvement, in that the gains in expressiveness are not outmatched by the increase in complexity of the semantics.

However, that would require working out exactly what the rules for multi-parameter type classes should be, especially:

1. How do we prevent ambiguity, where the user can declare instances that are equally valid for a given method call, and so resolution returns multiple results?

2. How can we prevent orphan instances?

This requires allocating some time to sit down, investigate prior art, and think deeply about how to get this right.

Then, if the resulting rules are simple enough, I can add this to the language.

# Implementation

Probably a major overhaul of both typeclass resolution and the storage and processing of declarations. On the other hand: an opportunity to improve these areas of the compiler.

**Comment by L-as** [comment]
The big orphan instance you want to avoid is (Haskell because I haven't memorised Austral syntax)
```haskell
instance C A b
instance C a B
```
since `C A B` then has two instances.
There was a Haskell RFC for preventing this: https://github.com/ghc-proposals/ghc-proposals/pull/279 (unclean solution IMO)
Rust _doesn't_ prevent you from defining the instances, but errs if you _have_ two conflicting instances (unclean because you can have two valid modules in isolation that cause an error if both are imported).
There is probably a good solution in the literature somewhere.

Do note, the above can be done without multi-parameter type classes by defining `instance C (T a B)` and `instance C (T A b)`, I'm not sure if Austral supports this (does it?). If it does, then that would also have to be prevented.

Naively, it seems like the issue is when you have a type variable that isn't wrapped in a local definition, so e.g. `instance C A (W b)` and `instance C (W a) B` would not conflict, `C A B` does not match any instance at all.

It's not clear to me whether it's possible to prevent ambiguity entirely.

**Reply by eudoxia0** [reply]
>The big orphan instance you want to avoid is (Haskell because I haven't memorised Austral syntax)

Yep, this is the central ambiguity. I think it can probably be prevented statically, but I need to think of the set of rules. 

C. A. R. Hoare's quote comes to mind: "There are two ways of constructing a software design: One way is to make it so simple that there are obviously no deficiencies, and the other way is to make it so complicated that there are no obvious deficiencies." This is an area where the strict and correct approach may not be simple to describe.

>Do note, the above can be done without multi-parameter type classes by defining instance `C (T a B)` and instance `C (T A b)`, I'm not sure if Austral supports this (does it?). If it does, then that would also have to be prevented.

Austral right now requires typeclass instance arguments to be either:

1. A concrete type.
2. A generic type applied to distinct type variables.

So you can't do `C (T a b)` or `C (T a a)`.
